File: python/lex.py
#Defining some tokens

EOF = 0
IDN = 1
INT = 2
CHAR = 3
STR = 4
OP = 5
OTHERS = 6

class lexer:
    def __init__(self, file_str):
        self.ptr = 0
        self.token = []
        self.token_type = 6
        self.file_str = file_str
    def get_token(self):
        return ''.join(self.token)
    def get_type(self):
        return self.token_type
    def next(self):
        self.token.clear()
        # Detecting whitespaces and just moving ahead
        while self.ptr < len(self.file_str) and self.file_str[self.ptr].isspace():
            self.ptr += 1
        # Detecting the end of file
        if self.ptr == len(self.file_str):
            self.token = None
            self.token_type = EOF
            print("Reached end of file! Bye")
            return None
        # Detecting variables
        if self.file_str[self.ptr].isalpha():
            self.token.append(self.file_str[self.ptr])
            self.ptr += 1
            while self.ptr < len(self.file_str) and self.file_str[self.ptr].isalnum() or self.file_str[self.ptr] == '_':
                self.token.append(self.file_str[self.ptr])
                self.ptr += 1
            self.token_type = IDN
            print(''.join(self.token))
            return ''.join(self.token)
        # Detecting integers
        if self.file_str[self.ptr].isnumeric():
            self.token.append(self.file_str[self.ptr])
            self.ptr += 1
            while self.ptr < len(self.file_str) and self.file_str[self.ptr].isnumeric():
                self.token.append(self.file_str[self.ptr])
                self.ptr += 1
            self.token_type = INT
            print(''.join(self.token))
            return ''.join(self.token)
        # Detecting Strings
        if self.file_str[self.ptr] == '\"' or  self.file_str[self.ptr] == '\'':
            # storing the start quotes to find the end quotes too
            terminator = self.file_str[self.ptr]
            self.token.append(self.file_str[self.ptr])
            self.ptr += 1
            while self.ptr < len(self.file_str) and self.file_str[self.ptr] != terminator:
                self.token.append(self.file_str[self.ptr])
                self.ptr += 1
            self.token.append(self.file_str[self.ptr])
            self.ptr += 1
            if terminator == "\"":
                self.token_type = STR
            else:
                self.token_type = CHAR
            print(''.join(self.token))
            return ''.join(self.token)
        # Detecting operators
        temp = self.file_str[self.ptr]
        if temp == '+' or temp == '-' or temp == '*' or temp == '/' or temp == '<' or temp == '>' or temp == '=' or temp == '&' or temp == '!':
            self.token.append(temp)
            terminator = temp
            self.ptr += 1
            if self.ptr < len(self.file_str) and self.file_str[self.ptr] == terminator or self.file_str[self.ptr] == "=":
                self.token.append(self.file_str[self.ptr])
                self.ptr += 1
            self.token_type = OP
            print(''.join(self.token))
            return ''.join(self.token)
        # Detecting Not Operator
        if temp == '!':
            self.token.append(temp)
            self.ptr += 1
            self.token_type = OP
            return ''.join(self.token)
        self.ptr += 1
        self.token = [temp]
        self.token_type = OTHERS
        print(''.join(temp))
        return ''.join(temp)

File: go_lang/compiler/lex.go
package compiler

import "unicode"

type Lex struct {
	ptr                                  int
	fileString                           string
	tokenType                            int
	token                                string
	EOF, IDN, INT, CHAR, STR, OP, OTHERS int
}

func (l Lex) initialize(fileString string) {
	l.ptr = 0
	l.token = ""
	l.tokenType = 6
	l.fileString = fileString
	l.EOF = 0
	l.IDN = 2
	l.CHAR = 3
	l.STR = 4
	l.OP = 5
	l.OTHERS = 6
}

func (l Lex) getToken() string {
	return l.token
}

func (l Lex) getType() int {
	return l.tokenType
}

func (l Lex) isAl(char byte) bool {
	if char >= 'a' && char <= 'z' || char >= 'A' && char <= 'Z' {
		return true
	}
	return false
}

func (l Lex) isNum(char byte) bool {
	if char >= '0' && char <= '9' {
		return true
	}
	return false
}

func (l Lex) isAlNum(char byte) bool {
	return l.isAl(char) || l.isNum(char)
}

func (l Lex) next() string {
	l.token = ""
	for l.ptr < len(l.fileString) && unicode.IsSpace(rune(l.fileString[l.ptr])) {
		l.ptr++
	}

	if l.ptr == len(l.fileString) {
		l.token = ""
		l.tokenType = l.EOF
		return ""
	}

	if l.isAl(l.fileString[l.ptr]) {
		l.token += string(l.fileString[l.ptr])
		l.ptr++
		for l.ptr < len(l.fileString) && l.isAlNum(l.fileString[l.ptr]) || l.fileString[l.ptr] == '_' {
			l.token += string(l.fileString[l.ptr])
			l.ptr++
		}
		l.tokenType = l.IDN
		return l.token
	}

	if l.isNum(l.fileString[l.ptr]) {
		l.token += string(l.fileString[l.ptr])
		l.ptr++
		for l.ptr < len(l.fileString) && l.isNum(l.fileString[l.ptr]) {
			l.token += string(l.fileString[l.ptr])
			l.ptr++
		}
		l.tokenType = l.INT
		return l.token
	}

	if l.fileString[l.ptr] == '"' || l.fileString[l.ptr] == '\'' {
		var terminator = l.fileString[l.ptr]
		l.token += string(l.fileString[l.ptr])
		l.ptr++
		for l.ptr < len(l.fileString) && l.fileString[l.ptr] != terminator {
			l.token += string(l.fileString[l.ptr])
			l.ptr++
		}
		l.token += string(l.fileString[l.ptr])
		l.ptr++
		if terminator == '"' {
			l.tokenType = l.STR
		}
		if terminator == '\'' {
			l.tokenType = l.CHAR
		}
		return l.token
	}
	var temp = l.fileString[l.ptr]
	if temp == '+' || temp == '-' || temp == '*' || temp == '/' || temp == '<' || temp == '>' || temp == '=' || temp == '&' || temp == '!' {
		l.token += string(temp)
		var terminator = temp
		l.ptr++
		if l.ptr < len(l.fileString) && l.fileString[l.ptr] == terminator || l.fileString[l.ptr] == '=' {
			l.token += string(l.fileString[l.ptr])
			l.ptr++
		}
		l.tokenType = l.OP
		return l.token
	}
	l.ptr++
	l.token += string(temp)
	l.tokenType = l.OTHERS
	return l.token
}
